src/libraries/System.Private.CoreLib/src/System/Globalization/InvariantModeCasing.cs

```diff
@@ -0,0 +1,346 @@
+                    if (source[i] != h || source[i + 1] != l)
+                    {
+                        break;
+                    }
+
+                    i += 2;
+                    continue;
+                }
+
+                if (ToUpper(source[i]) != source[i])
+                {
+                    break;
+                }
+
+                i++;
+            }
+
+            if (i >= s.Length)
+            {
+                return s;
+            }
+
+            return string.Create(s.Length, (s, i), static (destination, state) =>
+            {
+                ReadOnlySpan<char> src = state.s;
+                src.Slice(0, state.i).CopyTo(destination);
+                InvariantModeCasing.ToUpper(src.Slice(state.i), destination.Slice(state.i));
+            });
+        }
+
+        internal static void ToUpper(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(GlobalizationMode.Invariant);
+            Debug.Assert(source.Length <= destination.Length);
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                char c = source[i];
+                if (char.IsHighSurrogate(c) && i < source.Length - 1)
+                {
+                    char cl = source[i + 1];
+                    if (char.IsLowSurrogate(cl))
+                    {
+                        // well formed surrogates
+                        SurrogateCasing.ToUpper(c, cl, out char h, out char l);
+                        destination[i]   = h;
+                        destination[i+1] = l;
+                        i++; // skip the low surrogate
+                        continue;
+                    }
+                }
+
+                destination[i] = ToUpper(c);
+            }
+        }
+
+        internal static void ToLower(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(GlobalizationMode.Invariant);
+            Debug.Assert(source.Length <= destination.Length);
+
+            for (int i = 0; i < source.Length; i++)
+            {
+                char c = source[i];
+                if (char.IsHighSurrogate(c) && i < source.Length - 1 )
+                {
+                    char cl = source[i + 1];
+                    if (char.IsLowSurrogate(cl))
+                    {
+                        // well formed surrogates
+                        SurrogateCasing.ToLower(c, cl, out char h, out char l);
+                        destination[i]   = h;
+                        destination[i+1] = l;
+                        i++; // skip the low surrogate
+                        continue;
+                    }
+                }
+
+                destination[i] = ToLower(c);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static (uint, int) GetScalar(ref char charA, int index, int length)
+        {
+            if (!char.IsHighSurrogate(charA) || index >= length - 1)
```

Comment:
Time-of-check/time-of-use: since _charA_ is a reference, the value it points to can change between the `IsHighSurrogate` call here and the deref on lines 168 or 177 below. Similarly, _charB_ can change between the `IsLowSurrogate` call on line 172 and the deref on line 177.

We should dereference _charA_ exactly once into a local, then use that local for the remainder of the method instead of having multiple derefs. Similarly, we should deref _(ref charA + 1)_ exactly once (on line 171) into a local, then use that local for the remainder of the method instead of having multiple derefs.

Reply:
I have submitted the fix for that in the PR https://github.com/dotnet/runtime/pull/55771